cmd/scheduler/options/options.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package options
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/spf13/pflag"
+
+	"github.com/dapr/dapr/pkg/metrics"
+	"github.com/dapr/dapr/pkg/modes"
+	"github.com/dapr/dapr/pkg/security"
+	securityConsts "github.com/dapr/dapr/pkg/security/consts"
+	"github.com/dapr/kit/logger"
+)
+
+type Options struct {
+	Port                 int
+	HealthzPort          int
+	HealthzListenAddress string
+
+	ListenAddress    string
+	TLSEnabled       bool
+	TrustDomain      string
+	TrustAnchorsFile *string
+	SentryAddress    string
+	PlacementAddress string
+	Mode             string
+
+	ID                      string
+	ReplicaID               uint32
+	ReplicaCount            uint32
+	EtcdInitialPeers        []string
+	EtcdDataDir             string
+	EtcdClientPorts         []string
+	EtcdClientHTTPPorts     []string
+	EtcdSpaceQuota          int64
+	EtcdCompactionMode      string
+	EtcdCompactionRetention string
+
+	Logger  logger.Options
+	Metrics *metrics.Options
+
+	taFile string
+}
+
+func New(origArgs []string) (*Options, error) {
+	// We are using pflag to parse the CLI flags
+	// pflag is a drop-in replacement for the standard library's "flag" package, however…
+	// There's one key difference: with the stdlib's "flag" package, there are no short-hand options so options can be defined with a single slash (such as "daprd -mode").
+	// With pflag, single slashes are reserved for shorthands.
+	// So, we are doing this thing where we iterate through all args and double-up the slash if it's single
+	// This works *as long as* we don't start using shorthand flags (which haven't been in use so far).
+	args := make([]string, len(origArgs))
+	for i, a := range origArgs {
+		if len(a) > 2 && a[0] == '-' && a[1] != '-' {
+			args[i] = "-" + a
+		} else {
+			args[i] = a
+		}
+	}
+
+	var opts Options
+
+	// Create a flag set
+	fs := pflag.NewFlagSet("scheduler", pflag.ExitOnError)
+	fs.SortFlags = true
+
+	fs.IntVar(&opts.Port, "port", 50006, "The port for the scheduler server to listen on")
+	fs.IntVar(&opts.HealthzPort, "healthz-port", 8080, "The port for the healthz server to listen on")
+	fs.StringVar(&opts.HealthzListenAddress, "healthz-listen-address", "", "The listening address for the healthz server")
+
+	fs.StringVar(&opts.ListenAddress, "listen-address", "127.0.0.1", "The address for the Scheduler to listen on")
```

Comment:
`0.0.0.0`? 127.0.0.1 is not good in containers (aside from sidecars)

Reply:
I'm hesitant to put `0.0.0.0` as a default value due to the security implications.

Reply:
The default listen address for the other CP services is `0.0.0.0`. It won't work in Docker otherwise.

Reply:
[please see this PR with that change](https://github.com/cicoyle/dapr/pull/71/files#diff-0de3b642b59f384f86664dea2875d7602cdc27323f3220286b52ce0126c758f8)

Reply:
After further consideration, I left this as is since as a user I would not want anything listening on `0.0.0.0` by default bc if I forgot to change the default value then that is a security concern. Does anyone else feel strongly about this and could weigh in? cc/ @daixiang0 @yaron2 @artursouza @JoshVanL @mukundansundar

Reply:
Alessandro is correct that we need to listen to all addresses for this control plane to run from the container without having to change CLI to pass additional params (it would be a breaking change too for CLI to work with older Dapr versions or old CLI to work with newer Dapr versions).

All other control plane services (right now) listen on "" for this reason. I suggest we keep scheduler consistent (in this PR) and also listen on "". Then we have a separate PR to change all control planes (and the shared Dockerfile) to have them to listen to "" automatically from inside the docker image, without CLI having to pass additional params. We can have a new env var named DAPR_CONTROL_PLANE_LISTEN_ADDR or something that only control plane services will understand and it is set in Dockerfile globally.

Reply:
Software generally should not be made to listen on all interfaces and therefore be publicly available on the network by default. This is a clear security issue. Dapr services which are supported to be run from the CLI should default `localhost` for listening addresses. In the case of both kube and container target deployments, we have control over what arguments are given to the process via Helm and `dapr` CLI.

We should change `sentry`, `placement`, `daprd` and `scheduler` listen address defaults to `localhost`. I don't think there is a need for more config surface to be added because CLI flags already exist to change these values.

Reply:
I've updated the default for Scheduler to be `""` for consistency. Although, we should revisit changing the defaults separate of the Scheduler PR

Reply:
> Software generally should not be made to listen on all interfaces and therefore be publicly available on the network by default. 

While this is true for "regular" binaries, the difference is that Dapr control plane components are only meant to run inside Docker containers, even in self-hosted mode. We do not support running outside of a Docker container. Inside Docker, binding to 127.0.0.1 is not the desired behavior in almost all cases; if users want to make sure the container isn't exposed to the world, that should be done on Docker and not the binary.

Reply:
> Software generally should not be made to listen on all interfaces and therefore be publicly available on the network by default. This is a clear security issue. Dapr services which are supported to be run from the CLI should default `localhost` for listening addresses. In the case of both kube and container target deployments, we have control over what arguments are given to the process via Helm and `dapr` CLI.
> 
> We should change `sentry`, `placement`, `daprd` and `scheduler` listen address defaults to `localhost`. I don't think there is a need for more config surface to be added because CLI flags already exist to change these values.

Like I mentioned, requiring new params from CLI would be a breaking change. The CLI is meant to work with Dapr runtime version +-1, the container image should be done self-sufficient to run all binaries without extra params. I would not be in favor of breaking this compatibility if the cost is just a new env that users won't even need to set and come pre-set in the container image.

Anyway, for the purpose of this change, consistency is better than fixing for only one service and having to write on-off logic for scheduler.

Reply:
> > Software generally should not be made to listen on all interfaces and therefore be publicly available on the network by default.
> 
> While this is true for "regular" binaries, the difference is that Dapr control plane components are only meant to run inside Docker containers, even in self-hosted mode. We do not support running outside of a Docker container. Inside Docker, binding to 127.0.0.1 is not the desired behavior in almost all cases; if users want to make sure the container isn't exposed to the world, that should be done on Docker and not the binary.

Users can download the binaries and run it themselves in standalone mode without using Docker at all, but I agree that this would be rare and have not heard about it yet. Just a possible scenario.

Reply:
Moving the discussion on "if" and "how" to change control planes to listen to localhost to this issue: https://github.com/dapr/dapr/issues/7834

For the sake of this PR, can we all agree to keep consistency? 👍 or 👎

Reply:
I agree that it's a much better security posture to change to localhost across all control plane services, and that it should be done, outside the scope of this PR.